Panic when attempting to grep data read via fjson

```
$ curl -s -O https://data.gharchive.org/2023-02-08-3.json.gz &&
  gunzip 2023-02-08-3.json.gz &&
  head -53 2023-02-08-3.json > 53.json &&
  super -dynamic -vam -c "                                                                                                           
SELECT count() AS count
FROM '53.json' (format fjson)
WHERE grep('in case you have any feedback 😊', payload.pull_request.body);"

panic: (*vector.Fusion) 0x16280e7a01e0

goroutine 83 [running]:
github.com/brimdata/super/runtime/vam/expr.(*search).eval(0x16280e6a67e0, {0x16280e332670?, 0x16280ea80008?, 0x16280e606460?})
	/Users/phil/work/super/runtime/vam/expr/search.go:103 +0x634
github.com/brimdata/super/vector.Apply(0x1, 0x16280ec11d58, {0x16280e332670, 0x1, 0x1?})
	/Users/phil/work/super/vector/apply.go:25 +0xf6
github.com/brimdata/super/vector.Apply-range1(...)
	/Users/phil/work/super/vector/apply.go:30
github.com/brimdata/super/vector.Apply.rip.func1(...)
	/Users/phil/work/super/vector/apply.go:58
github.com/brimdata/super/vector.Apply(0x1, 0x16280ec11d58, {0x16280e332660, 0x1, 0x1?})
	/Users/phil/work/super/vector/apply.go:28 +0x207
github.com/brimdata/super/runtime/vam/expr.(*search).Eval(0x16280e6a67e0, {0xd749c18?, 0x16280e7a0420?})
	/Users/phil/work/super/runtime/vam/expr/search.go:61 +0x9e
github.com/brimdata/super/runtime/vam/op.(*Filter).Pull(0x16280e6a6810, 0x0)
	/Users/phil/work/super/runtime/vam/op/filter.go:26 +0x65
github.com/brimdata/super/runtime/vam/op/aggregate.(*scalarAggregate).Pull(0x16280e38fb90, 0x0)
	/Users/phil/work/super/runtime/vam/op/aggregate/scalar.go:48 +0xa9
github.com/brimdata/super/runtime/vam/op.(*combineParent).run(0x16280e84e680)
	/Users/phil/work/super/runtime/vam/op/combine.go:109 +0x3c
created by github.com/brimdata/super/runtime/vam/op.(*Combine).Pull.func1 in goroutine 1
	/Users/phil/work/super/runtime/vam/op/combine.go:42 +0x35
```

## Details

Repro is with super commit 6a4216d. The repro data is a subset of the GitHub Archive data set often used in testing.

It works fine if read with the regular JSON reader.

```
$ super -version
Version: v0.3.0-150-g6a4216db5

$ super -dynamic -vam -c "                                                                                                           
SELECT count() AS count
FROM '53.json'
WHERE grep('in case you have any feedback 😊', payload.pull_request.body);"

{count:0}
```

It also seems like there's something about the mix of the data in these 53 particular values that triggers it when read with fjson, as the problem is not triggered with only the first 52 values nor with only the 53rd value.

```
$ head -52 2023-02-08-3.json > 52.json

$ super -dynamic -vam -c "                                                                                                           
SELECT count() AS count
FROM '52.json' (format fjson)
WHERE grep('in case you have any feedback 😊', payload.pull_request.body);"

{count:0}

$ tail -1 53.json > 1.json

$ super -dynamic -vam -c "                                                                                                           
SELECT count() AS count
FROM '1.json' (format fjson)
WHERE grep('in case you have any feedback 😊', payload.pull_request.body);"

{count:0}
```

--- a/runtime/sam/expr/eval.go
+++ b/runtime/sam/expr/eval.go
@@ -213,7 +213,7 @@ func NewRegexpMatch(re *regexp.Regexp, e Evaluator) *RegexpMatch {
 }
 
 func (r *RegexpMatch) Eval(this super.Value) super.Value {
-	val := r.expr.Eval(this)
+	val := r.expr.Eval(this).Under()
 	switch id := val.Type().ID(); id {
 	case super.IDString:
 		if val.IsNull() {

--- a/runtime/sam/expr/fieldnamefinder.go
+++ b/runtime/sam/expr/fieldnamefinder.go
@@ -99,6 +99,8 @@ func (f *FieldNameMatcher) Match(typ super.Type) bool {
 		}
 	case *super.TypeError:
 		match = f.Match(typ.Type)
+	case *super.TypeFusion:
+		match = f.Match(typ.Type)
 	}
 	f.checkedIDs[id] = match
 	return match

--- a/runtime/vam/expr/search.go
+++ b/runtime/vam/expr/search.go
@@ -58,7 +58,11 @@ func NewSearchString(sctx *super.Context, s string, e Evaluator) Evaluator {
 }
 
 func (s *search) Eval(this vector.Any) vector.Any {
-	return vector.Apply(vector.ApplyRipUnions, s.eval, s.e.Eval(this))
+	return s.applyEval(s.e.Eval(this))
+}
+
+func (s *search) applyEval(vec vector.Any) vector.Any {
+	return vector.Apply(vector.ApplyRipUnions|vector.ApplyRipFusions, s.eval, vec)
 }
 
 func (s *search) eval(vecs ...vector.Any) vector.Any {
@@ -83,7 +87,7 @@ func (s *search) eval(vecs ...vector.Any) vector.Any {
 			if index != nil {
 				f = vector.Pick(f, index)
 			}
-			if vec2 := s.eval(f); vec2.Kind() != vector.KindNull {
+			if vec2 := s.applyEval(f); vec2.Kind() != vector.KindNull {
 				out = vector.Or(out, FlattenBool(vec2))
 			}
 		}
@@ -95,8 +99,6 @@ func (s *search) eval(vecs ...vector.Any) vector.Any {
 	case *vector.Map:
 		return vector.Or(s.evalForList(vec.Keys, vec.Offsets, index, n),
 			s.evalForList(vec.Values, vec.Offsets, index, n))
-	case *vector.Union:
-		return vector.Apply(vector.ApplyRipUnions, s.eval, vec)
 	case *vector.Error:
 		return s.eval(vec.Vals)
 	}
@@ -119,8 +121,8 @@ func (s *search) evalForList(vec vector.Any, offsets, index []uint32, length uin
 		for k := range n {
 			index2[k] = k + start
 		}
-		view := vector.Pick(vec, index2)
-		if FlattenBool(s.eval(view)).Bits.TrueCount() > 0 {
+		vec := s.applyEval(vector.Pick(vec, index2))
+		if FlattenBool(vec).Bits.TrueCount() > 0 {
 			out.Set(j)
 		}
 	}
@@ -157,7 +159,7 @@ func NewRegexpMatch(re *regexp.Regexp, e Evaluator) Evaluator {
 }
 
 func (r *regexpMatch) Eval(this vector.Any) vector.Any {
-	return vector.Apply(vector.ApplyRipUnions, r.eval, r.e.Eval(this))
+	return vector.Apply(vector.ApplyRipUnions|vector.ApplyRipFusions, r.eval, r.e.Eval(this))
 }
 
 func (r *regexpMatch) eval(vecs ...vector.Any) vector.Any {

--- a/runtime/ztests/expr/function/grep.yaml
+++ b/runtime/ztests/expr/function/grep.yaml
@@ -14,6 +14,7 @@ input: |
   {pattern:"z",input:{a:{b:"c"}}}
   {pattern:"a",input:"hello"::(string|null)}
   {pattern:"a",input:null::(string|null)}
+  {pattern:"c",input:{a:fusion({b?:"c",d?:_::string},<{b:string}>)}}
   {pattern:1,input:""}
   {pattern:null,input:"a"}
 
@@ -26,5 +27,6 @@ output: |
   [true,false]
   [true,false]
   [true,false]
+  [true,true]
   [error({message:"grep: pattern argument must be a string",on:1}),error({message:"grep: pattern argument must be a string",on:1})]
   [error({message:"grep: pattern argument must be a string",on:null}),error({message:"grep: pattern argument must be a string",on:null})]

--- a/runtime/ztests/expr/regexp.yaml
+++ b/runtime/ztests/expr/regexp.yaml
@@ -7,6 +7,7 @@ input: |
   "foobazbar"
   "xfoobazbar"
   "foobazbarx"
+  fusion("foolbar"::(null|string),<string>)
   127.0.0.1
   null
 
@@ -17,5 +18,6 @@ output: |
   true
   false
   false
+  true
   false
   null

--- a/runtime/ztests/expr/search-glob.yaml
+++ b/runtime/ztests/expr/search-glob.yaml
@@ -5,8 +5,10 @@ input: |
   {a:"foox",b:"there"}
   {a:"hello",b:"foox"}
   {a:"",b:"foo"}
+  fusion({a?:_::string,b?:"fool"},<{b:string}>)
 
 output: |
   {a:"foox",b:"there"}
   {a:"hello",b:"foox"}
   {a:"",b:"foo"}
+  fusion({a?:_::string,b?:"fool"},<{b:string}>)

--- a/runtime/ztests/expr/search-nested-field-regexp.yaml
+++ b/runtime/ztests/expr/search-nested-field-regexp.yaml
@@ -4,6 +4,7 @@ input: |
   {a:[{bar:"foo"}]}
   {a:[{car:"foo"}]}
   {a:[{c:"foo"},{b:1}]}
+  fusion({car:"foo"}::(null|{car:string}),<{car:string}>)
   null
   {a:[]::[{bar:null}]}
   {a:[]::[{b:null}]}
@@ -11,4 +12,5 @@ input: |
 output: |
   {a:[{bar:"foo"}]}
   {a:[{car:"foo"}]}
+  fusion({car:"foo"}::(null|{car:string}),<{car:string}>)
   {a:[]::[{bar:null}]}

--- a/runtime/ztests/expr/search-nested-field.yaml
+++ b/runtime/ztests/expr/search-nested-field.yaml
@@ -4,6 +4,8 @@ input: |
   {a:[{b:"foo"}]}
   {a:[{c:"foo"}]}
   {a:[{c:"foo"},{b:1}]}
+  fusion({a:1,b?:_::string},<{a:int64}>)
+  fusion({a:2,b?:"foo"},<{a:int64,b:string}>)
   null
   {a:[]::[{b:null}]}
   {a:[]::[{c:null}]}
@@ -11,4 +13,6 @@ input: |
 output: |
   {a:[{b:"foo"}]}
   {a:[{c:"foo"},{b:1}]}
+  fusion({a:1,b?:_::string},<{a:int64}>)
+  fusion({a:2,b?:"foo"},<{a:int64,b:string}>)
   {a:[]::[{b:null}]}

--- a/runtime/ztests/expr/search-primitives.yaml
+++ b/runtime/ztests/expr/search-primitives.yaml
@@ -4,7 +4,9 @@ input: |
   "foo"
   "bar"
   "foo"
+  fusion("foo"::(int64|string),<string>)
 
 output: |
   "foo"
   "foo"
+  fusion("foo"::(int64|string),<string>)

--- a/vector/fusion.go
+++ b/vector/fusion.go
@@ -47,8 +47,17 @@ func (f *Fusion) Serialize(b *scode.Builder, slot uint32) {
 }
 
 func Super(vec Any) Any {
-	if vec, ok := vec.(*Fusion); ok {
-		return vec.Values
+	if vec.Kind() == KindFusion {
+		var index []uint32
+		if view, ok := vec.(*View); ok {
+			index = view.Index
+			vec = view.Any
+		}
+		vals := vec.(*Fusion).Values
+		if index != nil {
+			vals = Pick(vals, index)
+		}
+		return vals
 	}
 	return vec
 }

--- a/walk.go
+++ b/walk.go
@@ -39,6 +39,9 @@ func Walk(typ Type, body scode.Bytes, visit Visitor) error {
 		return walkMap(typ, body, visit)
 	case *TypeError:
 		return Walk(typ.Type, body, visit)
+	case *TypeFusion:
+		ftyp, body := typ.DerefFusion(body)
+		return Walk(ftyp, body, visit)
 	}
 	return nil
 }

